backstory/src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java

```diff
@@ -0,0 +1,229 @@
+import com.google.cloud.vision.v1.AnnotateImageRequest;
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.BatchAnnotateImagesResponse;
+import com.google.cloud.vision.v1.EntityAnnotation;
+import com.google.cloud.vision.v1.Feature;
+import com.google.cloud.vision.v1.Image;
+import com.google.cloud.vision.v1.ImageAnnotatorClient;
+import com.google.cloud.vision.v1.ImageSource;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.protobuf.ByteString;
+import com.google.sps.servletData.AnalyzedImage;
+import java.io.ByteArrayOutputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet for managing image analysis with Vision API and blobstore for uploads */
+@WebServlet("/image-analysis")
+public class ImageAnalysisServlet extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Query to find all analyzed image entities.
+    Query query = new Query("analyzed-image");
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<AnalyzedImage> analyzedImages = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      String imageUrl = (String) entity.getProperty("imageUrl");
+      String labelsJsonArray = (String) ((Text) entity.getProperty("labelsJsonArray")).getValue();
+
+      analyzedImages.add(new AnalyzedImage(imageUrl, labelsJsonArray));
+    }
+
+    response.setContentType("application/json;");
+    Gson gson = new GsonBuilder().setPrettyPrinting().create();
+    String analyzedImagesJsonArray = gson.toJson(analyzedImages);
+    response.getWriter().println(analyzedImagesJsonArray);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Expecting a post request from Blobstore containing the data fields from the image-upload
+    // form. After having gone through Blobstore, the request will include the images uploaded. The
+    // form in the HTML will connect to the Blobstore URL, which encodes the images and then
+    // redirects the request to this Url. In this doPost, the images are analyzed with Vision API,
+    // and the analysis along with the image URL will be sent to datastore for permanent storage.
+
+    // Get the input from the form.
+    // Get the URL of the image that the user uploaded to Blobstore.
+    final String imageUrl = getUploadedFileUrl(request, "image-upload");
```

Comment:
Currently, the only way for imageUrl or bytes to be set to null is if the user uploaded a form with no image selected. I am updating the method so that, in this scenario, they are simply redirected back to the home page and nothing happens.

Reply:
What would happen if you respond with 400 (SC_ BAD_REQUEST)?
https://tomcat.apache.org/tomcat-5.5-doc/servletapi/javax/servlet/http/HttpServletResponse.html

Reply:
I set the servlet to validate these values, then return a 400 error if they fail.